### DISCUSSION TITLE
Insert Event Manually to force Usage Recalculation

### DISCUSSION THREAD

**Discussion by KuasarCloud** [apache/cloudstack#9597] [General]
Hi Community,

As discussed in (https://github.com/apache/cloudstack/discussions/9439) and reported in (https://github.com/apache/cloudstack/issues/9443), we’ve identified a situation where, if a public IP is released without first deleting a Load Balancer (LB) Rule, the system fails to generate the LB deletion event. As a result, the usage continues to charge for the LB Rule indefinitely.

Today, we encountered a similar issue where deleted volumes are still being charged in Usage, even though they no longer exist. We’re currently investigating the cause of this behavior and will report it as soon as we have more information.

In the meantime, I have a couple of questions:

Is it possible to manually add an entry in the event table and have the usage recalculated from a specific date?
If a project is suspended, does the usage continue to accumulate charges?
Looking forward to your insights.

Best regards,

**Comment by DaanHoogland** [comment]
@KuasarCloud you can add an event in the cloud.usage_event table. It will get copied to the cloud-usage.usage_event table and be process if the next job start at the period just before this event. This is quite tricky as old data is already in the tables. It is possible to delete old data in the intermediate table(s) and in the cloud_usage.cloud_usage table. In your case cloud_usage.usage_volume is important to check. in addition any events after your new event need to be set to unprocessed. your mileage may vary. simplest may be to just add the event for now and make it stop creating records in the future. Reconciling of usage is not well implemented.

**Comment by KuasarCloud** [comment]
Hey @DaanHoogland 

Thanks for your reply, is the any problem if create the id (sequential) and the uuid on the table event? by the way our management server, the database and the timezone un usage is set to America/Bogota, but the events on event tables have UTC, do you know how to change this behaviour?

**Reply by DaanHoogland** [reply]
You can create it but take care the right updates are done, the usage_event has an auto_increment id, so that is no problem. You can use UUID() in the insert statement for the uuid. Please test in a test environment instead of believing me! Even I don't.

As for the timezones, I know there is some functionality for this in, but not released yet: https://github.com/apache/cloudstack/pull/8243. When 4.20 is out you should upgrade to that version to get it.

**Comment by KuasarCloud** [comment]
Hi @DaanHoogland,

While checking on the `cloud_usage.usage_event` table for Load Balancer (LB) rules, I noticed that the LB events are copied, but the calculation seems to continue indefinitely. Below is an example where each LB rule element has its own `CREATE`/`DELETE` pair:

```sql
SELECT * FROM cloud_usage.usage_event 
WHERE type LIKE '%LB%' 
AND resource_id IN ('272','271','252','189','185','164');
```
| Type           | Created                            | Resource ID |
|------------|-------------------------|-------------|
| LB.CREATE | 2024-01-19 14:10:55     | 164               |
| LB.DELETE | 2024-01-22 20:58:40    | 164              |
| LB.CREATE | 2024-03-26 23:14:34    | 185              |
| LB.CREATE | 2024-04-03 16:07:47    | 189              |
| LB.DELETE | 2024-05-30 21:45:46    | 189              |
| LB.CREATE | 2024-07-25 20:59:47    | 252             |
| LB.CREATE | 2024-07-26 16:05:00    | 271             |
| LB.CREATE | 2024-07-26 16:13:55     | 272            |
| LB.DELETE | 2024-07-26 16:16:16     | 272             |
| LB.DELETE | 2024-07-26 17:49:17     | 252             |
| LB.DELETE | 2024-07-26 17:49:27     | 271             |
| LB.DELETE | 2024-07-27 18:27:51     | 185             |

---

The above results indicate that there are corresponding `CREATE` and `DELETE` events for the same `resource_id`. However, the usage calculation seems to be stuck in an indefinite loop.

Could you please provide guidance?

Thank you!

**Reply by DaanHoogland** [reply]
Can you find any other events for those resources ? (be sure to include resource type in your search)

**Comment by DaanHoogland** [comment]
@KuasarCloud , assuming this can be closed. please comment if you think otherwise.
